cpp/src/pca/pca_mg.cu

```diff
@@ -0,0 +1,513 @@
+                            Matrix::RankSizePair **rank_sizes, size_t n_parts,
+                            Matrix::Data<T> **trans_input, T *components,
+                            Matrix::Data<T> **input, T *singular_vals, T *mu,
+                            paramsPCAMG prms, bool verbose) {
+  int rank = handle.getImpl().getCommunicator().getRank();
+
+  std::vector<Matrix::RankSizePair *> ranksAndSizes(rank_sizes,
+                                                    rank_sizes + n_parts);
+  Matrix::PartDescriptor trans_desc(prms.n_rows, prms.n_components,
+                                    ranksAndSizes, rank);
+  std::vector<Matrix::Data<T> *> trans_data(trans_input, trans_input + n_parts);
+
+  std::vector<Matrix::Data<T> *> input_data(input, input + n_parts);
+
+  // TODO: These streams should come from cumlHandle
+  int n_streams = n_parts;
+  cudaStream_t streams[n_streams];
+  for (int i = 0; i < n_streams; i++) {
+    CUDA_CHECK(cudaStreamCreate(&streams[i]));
+  }
+
+  inverse_transform_impl(handle, trans_data, trans_desc, components, input_data,
+                         singular_vals, mu, prms, streams, n_streams, verbose);
+
+  for (int i = 0; i < n_streams; i++) {
+    CUDA_CHECK(cudaStreamSynchronize(streams[i]));
+  }
+
+  for (int i = 0; i < n_streams; i++) {
+    CUDA_CHECK(cudaStreamDestroy(streams[i]));
+  }
+}
+
+/**
+ * @brief performs MNMG fit and transform operation for the pca.
+ * @input param handle: the internal cuml handle object
+ * @input param rank_sizes: includes all the partition size information for the rank
+ * @input param n_parts: number of partitions
+ * @input param input: input data
+ * @output param trans_input: transformed input data
+ * @output param components: principal components of the input data
+ * @output param explained_var: explained var
+ * @output param explained_var_ratio: the explained var ratio
+ * @output param singular_vals: singular values of the data
+ * @output param mu: mean of every column in input
+ * @output param noise_vars: variance of the noise
+ * @input param prms: data structure that includes all the parameters from input size to algorithm
+ * @input param verbose
+ */
+template <typename T>
+void fit_transform_impl(cumlHandle &handle, Matrix::RankSizePair **rank_sizes,
+                        size_t n_parts, Matrix::Data<T> **input,
+                        Matrix::Data<T> **trans_input, T *components,
+                        T *explained_var, T *explained_var_ratio,
+                        T *singular_vals, T *mu, T *noise_vars,
+                        paramsPCAMG prms, bool verbose) {
+  int rank = handle.getImpl().getCommunicator().getRank();
+
+  std::vector<Matrix::RankSizePair *> ranksAndSizes(rank_sizes,
+                                                    rank_sizes + n_parts);
+  std::vector<Matrix::Data<T> *> input_data(input, input + n_parts);
+  Matrix::PartDescriptor input_desc(prms.n_rows, prms.n_cols, ranksAndSizes,
```

Comment:
It's strange this is being created in place and not being passed in from the Cython layer. We don't need to make the change in this PR, but we might consider creating an issue for this. Maybe many of these smaller refactor items can be combined in the same issue.

Reply:
Yeah, since my PR for this change originally was a big one, the cython layer API currently only passes it to the C++ functions that are actually called in cython

Reply:
Can you reference the issue here in todo? Will make it easier to track